Display number of items in queue in multiplayer

hwabis opened the pull request:
Closes https://github.com/ppy/osu/issues/17292 (@nagilson hope you don't mind if I try this)

![Capture](https://user-images.githubusercontent.com/62194990/158944402-ee63fdf2-cbda-4bf6-9339-8a620f5ee646.PNG)

The placement of the number is a little awkward? But I'm not sure how/where else to put it

hwabis committed:
display count of listQueue

peppy commented:
It should be part of the "Queue" text, as in "Queue (3)". Or something similar to the playlists queue display, with the count being a different colour:

![osu! 2022-03-18 at 05 54 32](https://user-images.githubusercontent.com/191335/158945568-7ec7709e-eadc-4a79-9484-16f8829de85a.png)

hwabis committed:
put count and tabs into new fillflowcontainer

hwabis commented:
![Capture](https://user-images.githubusercontent.com/62194990/159048377-555680c6-d2d6-414f-bb00-9c34d1990eb1.PNG)

I put the count and the OsuTabControl in a FillFlowContainer. (I wasn't sure how to put the count between "Queue" and "History", or change the text of "Queue" to reflect the count... so I just put it to the left for now.)

Also, the OsuTabControl goes too far right (visible with draw visualizer), so I will probably need to change its width

bdach commented:
> I wasn't sure how to put the count between "Queue" and "History", or change the text of "Queue" to reflect the count

You'd need to create a new class deriving from `OsuTabControl<MultiplayerPlaylistDisplayMode>` and override `CreateTabItem()` inside it to do that. Many other places do something like this, one of them is [here](https://github.com/ppy/osu/blob/master/osu.Game/Screens/Edit/Components/Menus/ScreenSelectionTabControl.cs).

hwabis committed:
(not working yet) add MultiplayerPlaylistTabControl, "Queue ({count})"

frenzibyte commented on osu.Game/Screens/OnlinePlay/Multiplayer/Match/Playlist/MultiplayerPlaylistTabControl.cs:
```diff
@@ -0,0 +1,34 @@
+{
+    public class MultiplayerPlaylistTabControl : OsuTabControl<MultiplayerPlaylistDisplayMode>
+    {
+        public int queueListCount;
```
You may want to have a look at `Bindable`s to share such data properly (this is 100% the reason why it's not working for you right now).

hwabis committed:
(still not working) try bindables instead of int

hwabis commented:
I'm having some trouble with this. Should `Text.Text` in `QueueTabItem` be updating automatically if `count`'s value is changed?

frenzibyte commented:
Correct, that's the reason why I'm suggesting to use `Bindable`s in the first place. See https://github.com/ppy/osu-framework/wiki/Bindable-Flow.

hwabis committed:
(works now) use BindValueChanged to set the Queue text

hwabis commented:
Ok, I think I got it working. Not sure if setting the entire `Text.Text` every time is the cleanest way though.

frenzibyte commented:
Requesting review is not necessarily required, we'll review this when we get the chance to.

hwabis committed:
make code prettier

frenzibyte commented on osu.Game/Screens/OnlinePlay/Multiplayer/Match/Playlist/MultiplayerPlaylist.cs:
```diff
@@ -70,6 +70,8 @@ protected override void LoadComplete()
         {
             base.LoadComplete();
             DisplayMode.BindValueChanged(onDisplayModeChanged, true);
+            queueList.Items.BindCollectionChanged(
+                (_, __) => playlistTabControl.queueListCount.Value = queueList.Items.Count, true);
```
This can be simplified by letting `QueueTabItem` listen for changes to the `queueList.Items` bindable directly rather than defining another bindable for that:
```csharp
public class MultiplayerPlaylistTabControl
{
    public readonly IBindableList<PlaylistItem> QueueItems = new BindableList<PlaylistItem>();
    
    ... (in QueueTabItem)
    
    QueueItems.BindCollectionChanged((_, __) => Text.Text = ..., true);
}

...

playlistTabControl.QueueItems.BindTarget = queueList.Items;
```

frenzibyte commented on osu.Game/Screens/OnlinePlay/Multiplayer/Match/Playlist/MultiplayerPlaylistTabControl.cs:
```diff
@@ -0,0 +1,35 @@
+{
+    public class MultiplayerPlaylistTabControl : OsuTabControl<MultiplayerPlaylistDisplayMode>
+    {
+        public Bindable<int> queueListCount = new Bindable<int>(0);
```
In general, bindables that receive value from external sources should be immutable (`IBindable...`) and readonly.

Also always instantiate the most derived version of any bindable to avoid potential issues from inheritance with base properties, i.e. use `BindableInt`/`BindableFloat`/`BindableDouble` instead of `Bindable<int>`/`Bindable<float>`/`Bindable<double>`:

And lastly, arguments that are equal to the default parameter values are redundant and will trigger CI.
```suggestion
        public readonly IBindable<int> queueListCount = new BindableInt();
```

frenzibyte commented on osu.Game/Screens/OnlinePlay/Multiplayer/Match/Playlist/MultiplayerPlaylistTabControl.cs:
```diff
@@ -0,0 +1,35 @@
+            public QueueTabItem(MultiplayerPlaylistDisplayMode value, Bindable<int> queueListCount)
+                : base(value)
+            {
+                queueListCount.BindValueChanged(
+                    _ => Text.Text = "Queue (" + queueListCount.Value + ")", true);
```
Listening to value changes should always happen at `LoadComplete` to avoid issues with the component itself not being loaded to take the value.

frenzibyte replied:
This should probably handle the case where there's no playlist items in the queue, by not displaying `(0)`.

frenzibyte commented on osu.Game/Screens/OnlinePlay/Multiplayer/Match/Playlist/MultiplayerPlaylistTabControl.cs:
```diff
@@ -0,0 +1,35 @@
+            public QueueTabItem(MultiplayerPlaylistDisplayMode value, Bindable<int> queueListCount)
+                : base(value)
+            {
+                queueListCount.BindValueChanged(
```
In all `Drawable`s, never listen to value changes from a bindable that's not stored locally on the drawable itself (that could lead to event leaks), define a local bindable and bind that to the extenral source to listen from it instead.

In other words, this is the recommended pattern (including using `LoadComplete` instead of constructor):
```csharp
new QueueTabItem { QueueListCount = { BindTarget = QueueListCount } }

...

public class QueueTabItem
{
    public readonly IBindable<int> QueueListCount = new BindableInt();

    protected override void LoadComplete()
    {
        base.LoadComplete();
        QueueListCount.BindValueChanged(v => Text.Text = $"Queue ({v.NewValue})", true);
    }
}
```

frenzibyte commented on osu.Game/Screens/OnlinePlay/Multiplayer/Match/Playlist/MultiplayerPlaylistTabControl.cs:
```diff
@@ -0,0 +1,35 @@
+
+        private class QueueTabItem : OsuTabItem
+        {
+            public QueueTabItem(MultiplayerPlaylistDisplayMode value, Bindable<int> queueListCount)
+                : base(value)
```
This is a `QueueTabItem`, so there's no need to require a `MultiplayerPlaylistDisplayMode` parameter:
```csharp
public QueueTabItem()
    : base(MultiplayerPlaylistDisplayMode.Queue)
{
}
```

frenzibyte reviewed (changes requested):
Thanks for your contribution.

Displaying the count as part of the queue tab item label itself seems fine for now, here's how that looks for reference:

![CleanShot 2022-03-19 at 07 13 34](https://user-images.githubusercontent.com/22781491/159106958-c937e9df-2a12-4b76-b447-56b52da75ea8.gif)

I've left a number of notes in regards to code standards and quality.

hwabis committed:
address frenzibyte review (refactor bindable flow, clean up CI stuff)

frenzibyte committed:
Bind queue items during playlist initialization

Along with where the other bindables are bound together (only bind, not
listening).

frenzibyte committed:
Simplify queue count text logic

frenzibyte commented:
I've pushed minor preliminary changes as well, looks pretty good. However, this is missing test coverage as this can easily regress in code. Check `TestSceneMultiplayerPlaylist` and `ChildrenOfType` usages for a guide on how to add proper tests for this.

hwabis committed:
add queue tab count test

hwabis commented:
I changed `OsuTabItem.Text` from protected to public, so I could compare it in the test. Not sure if that's good.

hwabis committed:
make test name clearer

frenzibyte commented on osu.Game.Tests/Visual/Multiplayer/TestSceneMultiplayerPlaylist.cs:
```diff
@@ -213,6 +233,17 @@ private void assertItemInHistoryListStep(int playlistItemId, int visualIndex)
+            string queueTabText = count > 0 ? $"Queue ({count})" : "Queue";
+            AddUntilStep($"Queue tab shows \"{queueTabText}\"", () =>
+            {
+                return this.ChildrenOfType<OsuTabControl<MultiplayerPlaylistDisplayMode>.OsuTabItem>()
+                           .First()
+                           .Text.Text.Equals(queueTabText);
```
You can check the text of the queue directly using another `ChildrenOfType` query:
```suggestion
                return this.ChildrenOfType<OsuTabControl<MultiplayerPlaylistDisplayMode>.OsuTabItem>()
                           .Single(t => t.Value == MultiplayerPlaylistDisplayMode.Queue)
                           .ChildrenOfType<OsuSpriteText>().Single().Text == queueTabText;
```
Never expose a property as `protected`/`public` for testing purposes if it's part of the hierarchy of that component, as you can use `ChildrenOfType` for that instead.

hwabis replied:
Whoops, this definitely makes more sense.

hwabis committed:
Revert OsuTabItem Text to protected, update tab count test

frenzibyte committed:
Merge branch 'master' into queue-list-counter
